tuning/trainercontroller/controllermetrics/metrics.py

```diff
@@ -0,0 +1,218 @@
+        """Computes the controller-metric (step-loss over window) and exposes the values of the variables used by the rules.
+
+        Args:
+            training_state: TrainerState object
+            training_args: TrainingArguments object
+            metrics: [optional] metrics data
+
+        Returns:
+            dict
+        """
+        # Compute the metric using the training state
+        loss_values = [l['loss'] for l in training_state.log_history if 'loss' in l]
+        n = len(loss_values)
+        if n <= self.__window_size:
+            return None
+        window = loss_values[n-self.__window_size:n]
+        consistently_increasing = True
+        for i in range(len(window)-1):
+            if window[i] > window[i+1]:
+                consistently_increasing = False
+                break
+        w = np.array(window)
+        avg_loss = np.mean(w)
+        std_loss = np.std(w, dtype=np.float64)
+        exposed_data = {'consistently_increasing': int(consistently_increasing), \
+                    'average_loss': avg_loss,\
+                    'std_loss': std_loss,\
+                    'window': window}
+        return exposed_data
+
+class EpochLoss(MetricHandler):
+    """Implements the controller metric which evaluates loss-per-epoch over a user-defined window"""
+
+    def __init__(self, window_size):
+        # Initialize the handler arguments
+        self.__window_size = window_size
+        self.__cache = []
+        
+    def validate(self, training_args):
+        """Validate the training arguments (e.g logging_steps) are compatible with the computation of this metric
+
+        Args:
+            training_args: Training arguments
+
+        Returns:
+            bool
+        """
+        return training_args.logging_strategy == 'steps' and \
+            training_args.logging_steps == 1
+
+    def __externalize_data(self):
+        """Create the dictionary of exposed variables used by rules.
+
+        Returns:
+            dict
+        """
+        if len(self.__cache) < self.__window_size:
+            return None
+        exposed_data = {}
+        for elem in reversed(self.__cache):
+            for k, v in elem.items():
+                l = []
+                if k in exposed_data:
+                    l = exposed_data[k]
+                l.append(v)
+                exposed_data[k] = l
+        return exposed_data
+
+    def compute(self, training_state, training_args=None, metrics=None):
+        """Computes the controller-metric (epoch-loss over window) and exposes the values of the variables used by the rules.
+
+        Args:
+            training_state: TrainerState object
+            training_args: TrainingArguments object
+            metrics: [optional] metrics data
+
+        Returns:
+            dict
+        """
+        previous_epoch = -1
+        loss_array = None
+        logs_latest_first = list(reversed(training_state.log_history))
```

Comment:
Instead of processing the entire history, reversing and more, can we instead just store the index to which we have previously processed as a object variable and process only the remaining?